fv/containers/containers.go

```diff
@@ -312,10 +314,56 @@ func (c *Container) copyOutputToLog(streamName string, stream io.Reader, done *s
 	defer done.Done()
 	scanner := bufio.NewScanner(stream)
 	scanner.Buffer(nil, 10*1024*1024) // Increase maximum buffer size (but don't pre-alloc).
+
+	// Felix is configured with the race detector enabled. When the race detector fires, we get output like this:
+	//
+	// ==================
+	// WARNING: DATA RACE
+	// <stack trace>
+	// ==================
+	//
+	// We capture that output and emit it to a dedicated log file so that the CI job can save it off.
+	// foundDataRace is set to true when we see the WARNING line and then it is set back to false when we
+	// see the trailing "==================".  We collect the text of the warning in dataRaceText.
+	//
+	// We do this for all containers because we already have the machinery here.
+	foundDataRace := false
+	dataRaceText := ""
+	dataRaceFile, err := os.OpenFile("data-races.log", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		log.WithError(err).Error("Failed to open data race log file.")
+	}
+	defer func() {
+		err := dataRaceFile.Close()
+		Expect(err).NotTo(HaveOccurred(), "Failed to write to data race log (close).")
+	}()
+
 	for scanner.Scan() {
 		line := scanner.Text()
 		log.Info(c.Name, "[", streamName, "] ", line)
 
+		// Capture data race warnings and log to file.
+		if strings.Contains(line, "WARNING: DATA RACE") {
+			_, err := fmt.Fprintf(dataRaceFile, "Detected data race while running test: %s\n", ginkgo.CurrentGinkgoTestDescription().FullTestText)
+			Expect(err).NotTo(HaveOccurred(), "Failed to write to data race log.")
+			foundDataRace = true
+		}
+		if foundDataRace {
+			var err error
+			if strings.Contains(line, "==================") {
+				foundDataRace = false
+				c.mutex.Lock()
+				c.dataRaces = append(c.dataRaces, dataRaceText)
+				c.mutex.Unlock()
+				dataRaceText = ""
+				_, err = dataRaceFile.WriteString("\n\n")
+			} else {
+				dataRaceText += line + "\n"
+				_, err = dataRaceFile.WriteString(line + "\n")
```

Comment:
You're right, the logs could overlap but I'd rather not add a mutex just in case we don't get the `=======` line for some reason (since we'd then deadlock without adding more machinery). In practice, I haven't seen overlaps so I'm inclined to leave it as is.